elixir/apps/domain/lib/domain/flows.ex

```diff
@@ -49,40 +49,96 @@ defmodule Domain.Flows do
+
+      Flow.Query.by_id(id)
+      |> Authorizer.for_subject(Flow, subject)
+      |> Repo.fetch()
+      |> case do
+        {:ok, resource} -> {:ok, Repo.preload(resource, preload)}
+        {:error, reason} -> {:error, reason}
+      end
+    else
+      false -> {:error, :not_found}
+      other -> other
+    end
+  end
+
   def list_flows_for(assoc, subject, opts \\ [])
 
   def list_flows_for(%Policies.Policy{} = policy, %Auth.Subject{} = subject, opts) do
     Flow.Query.by_policy_id(policy.id)
-    |> list(subject, opts)
+    |> list_flows(subject, opts)
   end
 
   def list_flows_for(%Resources.Resource{} = resource, %Auth.Subject{} = subject, opts) do
     Flow.Query.by_resource_id(resource.id)
-    |> list(subject, opts)
+    |> list_flows(subject, opts)
   end
 
   def list_flows_for(%Clients.Client{} = client, %Auth.Subject{} = subject, opts) do
     Flow.Query.by_client_id(client.id)
-    |> list(subject, opts)
+    |> list_flows(subject, opts)
   end
 
   def list_flows_for(%Gateways.Gateway{} = gateway, %Auth.Subject{} = subject, opts) do
     Flow.Query.by_gateway_id(gateway.id)
-    |> list(subject, opts)
+    |> list_flows(subject, opts)
   end
 
-  defp list(queryable, subject, opts) do
+  defp list_flows(queryable, subject, opts) do
     with :ok <- Auth.ensure_has_permissions(subject, Authorizer.view_flows_permission()) do
       {preload, _opts} = Keyword.pop(opts, :preload, [])
 
       {:ok, flows} =
         queryable
-        |> Authorizer.for_subject(subject)
+        |> Authorizer.for_subject(Flow, subject)
         |> Ecto.Query.order_by([flows: flows], desc: flows.inserted_at, desc: flows.id)
         |> Ecto.Query.limit(50)
         |> Repo.list()
 
       {:ok, Repo.preload(flows, preload)}
     end
   end
+
+  def upsert_activities(activities) do
+    {num, _} =
+      Repo.insert_all(Activity, activities, on_conflict: :nothing)
+
+    {:ok, num}
+  end
+
+  def list_flow_activities_for(
+        %Flow{} = flow,
+        ended_after,
+        started_before,
+        %Auth.Subject{} = subject
+      ) do
+    Activity.Query.by_flow_id(flow.id)
+    |> list_activities(ended_after, started_before, subject)
+  end
+
+  def list_flow_activities_for(
+        %Accounts.Account{} = account,
+        ended_after,
+        started_before,
+        %Auth.Subject{} = subject
+      ) do
+    Activity.Query.by_account_id(account.id)
+    |> list_activities(ended_after, started_before, subject)
+  end
+
+  defp list_activities(queryable, ended_after, started_before, subject) do
+    with :ok <- Auth.ensure_has_permissions(subject, Authorizer.view_flows_permission()) do
+      queryable
+      |> Activity.Query.by_window_ended_at({:greater_than, ended_after})
+      |> Activity.Query.by_window_started_at({:less_than, started_before})
+      |> Authorizer.for_subject(Activity, subject)
+      |> Ecto.Query.order_by([activities: activities], asc: activities.window_started_at)
```

Comment:
Is there any reason for us to rename entities in Elixir because of other platform terminology? If it's for grep-ing you can always do it in `kotlin/` subfolder, I do the same for `elixir/` all the time because a lot of things are named similarly everywhere.

`Activity` is something happening within the `Flow`. Basically, it's a `Metric`, but `Metric` term is overloaded and expected to be related to DevOps stuff rather than client network activity on a gateway. So naming them `NetFlows` or `AuthorizedFlows` would not reflect what they mean within `Flows` context that already has a `Flow`.

I was thinking of calling `Flow` a `NetworkFlow` initially but we also have `Network` context already for other stuff, so I felt that it would be ambiguous. `AuthorizedFlows` - I would avoid anything `Auth*` because it's also an extremely overloaded term and I don't feel like we won't use this term internally (eg. on a standup) - we will call them "flows" while speaking so that's how it should be named.